src/preact/base-element.js

```diff
@@ -0,0 +1,314 @@
+      const deferred =
+        this.scheduledRenderDeferred_ ||
+        (this.scheduledRenderDeferred_ = new Deferred());
+      this.context_.renderable = true;
+      this.context_.playable = true;
+      this.scheduleRender_();
+      return deferred.promise;
+    }
+
+    /** @override */
+    mutatedAttributesCallback() {
+      if (this.container_) {
+        this.scheduleRender_();
+      }
+    }
+
+    /** @private */
+    scheduleRender_() {
+      if (!this.scheduledRender_) {
+        this.scheduledRender_ = true;
+        this.mutateElement(this.boundRerender_);
+      }
+    }
+
+    /** @private */
+    unmount_() {
+      this.mounted_ = false;
+      if (this.container_) {
+        render(Fragment, this.container_);
+      }
+    }
+
+    /** @private */
+    rerender_() {
+      // If the component unmounted before the scheduled render runs, exit
+      // early.
+      if (!this.mounted_) {
+        return;
+      }
+
+      if (!this.container_) {
+        if (opts.children || opts.passthrough) {
+          this.container_ = this.element.attachShadow({mode: 'open'});
+        } else {
+          const container = this.win.document.createElement('i-amphtml-c');
+          this.container_ = container;
+          this.applyFillContent(container);
+          this.element.appendChild(container);
+        }
+      }
+
+      const props = collectProps(
+        this.element,
+        devAssert(opts),
+        this.defaultProps_
+      );
+
+      // While this "creates" a new element, diffing will not create a second
+      // instance of Component. Instead, the existing one already rendered into
+      // this element will be reused.
+      const cv = createElement(Component, props);
+
+      const v = createElement(withAmpContext, this.context_, cv);
+
+      render(v, this.container_);
+
+      const deferred = this.scheduledRenderDeferred_;
+      if (deferred) {
+        deferred.resolve();
+        this.scheduledRenderDeferred_ = null;
+      }
+    }
+  };
+}
+
+/**
+ * @param {!AmpElement} element
+ * @param {!AmpElementOptions} opts
+ * @param {!JsonObject|null|undefined} defaultProps
+ * @return {!JsonObject}
+ */
+function collectProps(element, opts, defaultProps) {
+  const props = /** @type {!JsonObject} */ ({...defaultProps});
+
+  // Class.
+  if (opts.className) {
+    props['className'] = opts.className;
+  }
+
+  // Props.
+  const defs = opts.props || {};
+  for (const name in defs) {
+    const def = defs[name];
+    const value = element.getAttribute(def.attr);
+    if (value == null) {
+      props[name] = def.default;
+    } else {
+      const v =
+        def.type == 'number'
+          ? Number(value)
+          : def.type == 'Element'
+          ? // TBD: what's the best way for element referencing compat between
+            // React and AMP? Currently modeled as a Ref.
+            {current: element.getRootNode().getElementById(value)}
+          : value;
+      props[name] = v;
+    }
```

Comment:
This nested ternary seems strange to me.
I'd love if we could simplify the control flow.

What do you think about flattening them with a series of else-ifs?
```js
if (value == null) {
  props[name] = def.default;
} else if (def.type == 'number') {
  props[name] = Number(value)
} else if (def.type == 'Element') {
  // TBD: what's the best way for element referencing compat between
  // React and AMP? Currently modeled as a Ref.
  props[name] = {current: element.getRootNode().getElementById(value)}
} else {
  props[name] = value;
}
 ```